n2survey/lime.py

```diff
@@ -0,0 +1,436 @@
+        list[tuple[str, str]]: List of pairs (<subsection varName>,
+          <subsection cleaned label>)
+    """
+    # TODO: Add validation that varName is provided
+    # TODO: Add validation that there is only one <text> tag
+    return [
+        (subquestion["varName"], _get_clean_string(subquestion.find("text")))
+        for subquestion in question.find_all("subQuestion", recursive=False)
+    ]
+
+
+def _parse_single_question_response(response: Tag) -> tuple[dict, Optional[dict]]:
+    """Parse single <response> element of a question
+
+    Args:
+        response (Tag): bs4 tag of <response> element
+
+    Raises:
+        AssertionError: Unknown response type. First child must be
+          either "free" or "fixed".
+
+    Returns:
+        tuple[dict, Optional[dict]]: Pair:
+          1. Main response data: name, format, length, label, choices
+          2. Contingent question data. None if there is no contingent
+          question. Contains, name, text, length, format, and:
+            * contingent_of_name - name of the parent <response> element
+            * contingent_of_choice - <value> of the parent <choice> element
+    """
+    # Common response structure
+    parsed_response = {
+        "name": response["varName"],
+        "format": None,
+        "length": None,
+        "label": None,
+        "choices": None,
+    }
+    contingent_response = None
+
+    # Get first child node of <response> section
+    response_data = response.findChild()
+
+    # Parse non-fixed question
+    if response_data.name == "free":
+        parsed_response.update(
+            format=_get_clean_string(response_data.format),
+            length=_get_clean_string(response_data.length),
+            label=_get_clean_string(response_data.label),
+        )
+    # Parse fixed question (i.e. with choises)
+    elif response_data.name == "fixed":
+        # Parse choices
+        choices = {
+            _get_clean_string(category.value): _get_clean_string(category.label)
+            for category in response_data.find_all("category", recursive=False)
+        }
+        parsed_response.update(choices=choices)
+
+        # Parse contingent question
+        contingent_question = response_data.find("contingentQuestion")
+        if contingent_question is not None:
+            assert (
+                len(response_data.find_all("contingentQuestion")) == 1
+            ), f"Too many 'contingentQuestion's for response {response}"
+
+            contingent_response = {
+                "name": contingent_question["varName"],
+                "text": _get_clean_string(contingent_question.find("text")),
+                "length": _get_clean_string(contingent_question.length),
+                "format": _get_clean_string(contingent_question.format),
+                "contingent_of_name": parsed_response["name"],
+                "contingent_of_choice": _get_clean_string(
+                    contingent_question.parent.value
+                ),
```

Comment:
the "longtext other" type question under the tag structure as: "<response> ->  <fixed> -> <category> -> <contingentQuestion>", if you have already add the <response> "varName" to the "parsed_response" dict, which means it will add a {"name": "#T", "format":...}  to it, and here continue parse this <contingentQuestion>, it will add another {"name": "#other", "format":...} to the "parsed_response" , which means you would add this "longtext other" type question twice.

Reply:
I am not sure if I got your question... Could you please take a look at `test_choice_question_with_contingent` there is an example of the output in case of contingent question.
Please note that here we do not add it to `parsed_response`, it is `contingent_response`. And the output of the function is a tuple `(parsed_response, contingent_response)`